lexxpluss_apps/src/board_controller.cpp

```diff
@@ -1144,6 +1145,17 @@ class state_controller { // Variables Implemented
     bool is_esw_asserted(){
         return esw.is_asserted();
     }
+    bool is_emergency() const {
+        bool rtn{false};
+        if (state != POWER_STATE::OFF) {
+            rtn = esw.is_asserted() ||
+               bsw.is_asserted() ||
+               state == POWER_STATE::SUSPEND ||
+               state == POWER_STATE::RESUME_WAIT ||
+               mbd.emergency_stop_from_ros();
```

Comment:
I thought esw.is_asserted() could be removed.

It may be outside the scope of this change, but it might be better to switch the state to SUSPEND when bsw.is_asserted() and mbd.emergency_stop_from_ros() are true.

---

esw.is_asserted()は削除できるのではないかと思いました。

今回の変更の範囲外かもしれませんが、bsw.is_asserted()、mbd.emergency_stop_from_ros()がtrueの時は状態をSUSPENDに切り替える方が良いかもしれません。

Reply:
レビューありがとうございます！！

> esw.is_asserted()は削除できるのではないかと思いました。

esw.is_asserted()を削除した場合、起動時や手動充電中にemergency switchを押下してもLEDの赤色点滅がされなくなってしまいます！！

> 今回の変更の範囲外かもしれませんが、bsw.is_asserted()、mbd.emergency_stop_from_ros()がtrueの時は状態をSUSPENDに切り替える方が良いかもしれません。

mbd.emergency_stop_from_ros()に関しては以下の分岐にてSUSPENDに切り替える処理が実行されています！！bsw.is_assertedに関しては、私も同意見なのですが現状の実装に合わせるためこのようにしております。

https://github.com/LexxPluss/LexxHard-SensorControlBoard-Firmware/blob/28180bd24928937f6f0587f80916e4a3f04ddcd8/lexxpluss_apps/src/board_controller.cpp#L1307

---

Thanks for your reviewing!!

> I thought esw.is_asserted() could be removed.

I guess that If esw.is_asserted() is removed、LED doesn't blink with red in launching or manual charging even if emergency switch is pressed.

> It may be outside the scope of this change, but it might be better to switch the state to SUSPEND when bsw.is_asserted() and mbd.emergency_stop_from_ros() are true.

mbd.emergency_stop_from_ros() is already handled in the following. I agree to the idea of handling bsw.is_asserted. But These codes are motivated to to keep the current behavior.

https://github.com/LexxPluss/LexxHard-SensorControlBoard-Firmware/blob/28180bd24928937f6f0587f80916e4a3f04ddcd8/lexxpluss_apps/src/board_controller.cpp#L1307

Reply:
> esw.is_asserted()を削除した場合、起動時や手動充電中にemergency switchを押下してもLEDの赤色点滅がされなくなってしまいます！！

なるほど。単に削除はできないですね。

> mbd.emergency_stop_from_ros()に関しては以下の分岐にてSUSPENDに切り替える処理が実行されています！！bsw.is_assertedに関しては、私も同意見なのですが現状の実装に合わせるためこのようにしております。

mbd.emergency_stop_from_ros() はここの条件式からは取り除けそうですね。
bswは1秒だけとか限定的だから状態を切り替えないような判断がどこかであったのかもしれないですね。

----

> I guess that If esw.is_asserted() is removed、LED doesn't blink with red in launching or manual charging even if emergency switch is pressed.

I see. We can't simply remove it.

> mbd.emergency_stop_from_ros() is already handled in the following. I agree to the idea of handling bsw.is_asserted. But These codes are motivated to to keep the current behavior.

I think mbd.emergency_stop_from_ros() could be removed from the conditional expression here.
Maybe there was a decision somewhere not to switch states because bsw is limited to 1 second or something.